salt/_modules/better_kubernetes.py

```diff
@@ -0,0 +1,148 @@
+"""Execution methods for management of Kubernetes objects.
+
+This module relies on the `kubernetes` custom Salt utils module for parsing
+K8s object manifests, and providing direct bindings to the Python `kubernetes`
+client.
+
+TODO:
+- handle custom objects
+- add some logging for debug purposes
+- add some documentation for each method
+- support giving a `path` to a file instead of directly the `manifest` data
+- improve management of `kubeconfig` and `context` parameters, relying on
+  master configuration and sane defaults - look into `__opts__`
+"""
+from salt.exceptions import CommandExecutionError
+
+try:
+    import kubernetes.client as k8s_client
+except ImportError:
+    HAS_LIBS = False
+else:
+    HAS_LIBS = True
+
+__virtualname__ = 'metalk8s_kubernetes'
+
+
+def __virtual__():
+    if not HAS_LIBS:
+        return False, 'Missing Python `kubernetes.client` dependency'
+
+    return __virtualname__
+
+
+def _extract_obj_and_details(manifest):
+    try:
+        details = __utils__['kubernetes.get_object_details'](manifest)
+        obj = __utils__['kubernetes.convert_manifest_into_object'](manifest)
+    except ValueError as exc:
+        raise CommandExecutionError('Invalid manifest: {!s}'.format(exc))
+
+    return obj, details
+
+
+def _extract_kwargs(obj, details):
+    kwargs = {'name': obj.metadata.name}
+
+    if details.scope == 'namespaced':
+        kwargs['namespace'] = obj.metadata.namespace
+
+    if hasattr(details, 'group'):
+        # This object represents a CustomResource
```

Comment:
Why? All K8s objects belong to a 'group'?

Reply:
Yeah, that part doesn't look nice to me... However, that's one specificity from the Python K8s client when it comes to managing custom objects and the signature of the client methods.

For instance, a _standard_ object would be created with [`CoreV1Api.create_namespaced_pod(namespace, body)`](https://github.com/kubernetes-client/python/blob/master/kubernetes/docs/CoreV1Api.md#create_namespaced_pod).

On the other hand, custom objects need [`CustomObjectsApi.create_namespaced_custom_object(group, version, namespace, plural, body)`](https://github.com/kubernetes-client/python/blob/master/kubernetes/docs/CustomObjectsApi.md#create_namespaced_custom_object).

Now that I look again at those examples, it seems I made another unrelated mistake: `create` methods don't accept a `name` argument...
# Custom objects
kubernetes.client.apis.CustomObjectsApi.create_namespaced_custom_object(

Reply:
However, I _could_ inject these arguments when construction the `CRDetails` object. :+1: this comment if that would suit you :)

Reply:
I went for the injection of special params when manipulating custom objects, for simplicity on the caller side.